lib/sorbet-rails/model_plugins/active_record_assoc.rb

```diff
@@ -19,22 +19,31 @@ def generate(root)
     model_class_rbi.create_include(assoc_module_name)
 
     @model_class.reflections.sort.each do |assoc_name, reflection|
+      begin
+        # TODO allow people to specify the possible values of polymorphic associations
+        assoc_class = reflection.polymorphic? ? "T.untyped" : "::#{reflection.klass.name}"
```

Comment:
Playing around with this more, I think this is the most robust way to handle this problem. Our main goal is to always know the type if possible and ideally our solution for doing so would be resilient to changes within Rails (e.g. implementation changes, new options added, etc). Consider the following:
```ruby
has_many :pets, through: :pet_ownerships, class_name: 'Dog', source: :baz, source_type: 'Xyz'
```

Different permutations of those options will yield a different class that Rails will use for this association. In Rails 6 it looks for the class in the following order:
1. `class_name:` (https://github.com/rails/rails/blob/master/activerecord/lib/active_record/reflection.rb#L166)
2. `source_type:` (https://github.com/rails/rails/blob/master/activerecord/lib/active_record/reflection.rb#L1001-L1004)
3. `source_reflection.class_name` (same link)
    - I think this may have been the issue you ran into. If this were a reflection where `source` was polymorphic and it didn't specify either the `class_name` or `source_type` option then it would fallback to trying to call `source.klass.name` which would blow up because it couldn't know the type

After this digging, we _could_ apply that logic to the existing checks but it feels a little bad reaching into the implementation details of `reflection.klass` to know if it's safe to call it vs just calling it and rescuing.

Thoughts?

Reply:
I think there is something about the "through:" connection that we tried to handle with the recursive lookup that this doesn't handle. I'll take another pass later to understand what we tried to do there.

Reply:
> After digging into this we could apply that logic to the existing checks but it feels a little bad reaching into the implementation details of reflection.klass to know if it's safe to call it vs just calling it and rescuing.

I understand the rationale now. It could be a better way of handling it than recursing through the relation chain (and maybe faster). At the end of the day, we would trust `reflection.klass` anyway. I'll investigate that in a follow up (if I can find the original issue).